lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx

```diff
@@ -0,0 +1,542 @@
+#include "openturns/ComposedDistribution.hxx"
+#include "openturns/Uniform.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(EfficientGlobalOptimization);
+
+static const Factory<EfficientGlobalOptimization> Factory_EfficientGlobalOptimization;
+
+/* Constructor with parameters */
+EfficientGlobalOptimization::EfficientGlobalOptimization()
+  : OptimizationAlgorithmImplementation()
+  , solver_(new Cobyla)
+  , useDefaultSolver_(true)
+  , multiStartExperimentSize_(ResourceMap::GetAsUnsignedInteger("EfficientGlobalOptimization-DefaultMultiStartExperimentSize"))
+  , multiStartNumber_(ResourceMap::GetAsUnsignedInteger("EfficientGlobalOptimization-DefaultMultiStartNumber"))
+  , parameterEstimationPeriod_(ResourceMap::GetAsUnsignedInteger("EfficientGlobalOptimization-DefaultParameterEstimationPeriod"))
+  , improvementFactor_(1.0)
+  , correlationLengthFactor_(1.0)
+  , aeiTradeoff_(1.0)
+{
+}
+
+/* Constructor with parameters */
+EfficientGlobalOptimization::EfficientGlobalOptimization(const OptimizationProblem & problem,
+                                                         const KrigingResult & krigingResult)
+  : OptimizationAlgorithmImplementation(problem)
+  , krigingResult_(krigingResult)
+  , solver_(new Cobyla)
+  , useDefaultSolver_(true)
+  , multiStartExperimentSize_(ResourceMap::GetAsUnsignedInteger("EfficientGlobalOptimization-DefaultMultiStartExperimentSize"))
+  , multiStartNumber_(ResourceMap::GetAsUnsignedInteger("EfficientGlobalOptimization-DefaultMultiStartNumber"))
+  , parameterEstimationPeriod_(ResourceMap::GetAsUnsignedInteger("EfficientGlobalOptimization-DefaultParameterEstimationPeriod"))
+  , improvementFactor_(1.0)
+  , correlationLengthFactor_(1.0)
+  , aeiTradeoff_(1.0)
+{
+  checkProblem(problem);
+}
+
+
+class ExpectedImprovementEvaluation : public NumericalMathEvaluationImplementation
+{
+public:
+  ExpectedImprovementEvaluation (const NumericalScalar optimalValue,
+                                      const KrigingResult & metaModelResult,
+                                      const NumericalMathFunction & noiseModel)
+  : NumericalMathEvaluationImplementation()
+  , optimalValue_(optimalValue)
+  , metaModelResult_(metaModelResult)
+  , noiseModel_(noiseModel)
+  {
+  }
+
+  virtual ExpectedImprovementEvaluation * clone() const
+  {
+    return new ExpectedImprovementEvaluation(*this);
+  }
+
+  NumericalPoint operator()(const NumericalPoint & x) const
+  {
+    const NumericalScalar mx = metaModelResult_.getMetaModel()(x)[0];
+    const NumericalScalar fmMk = optimalValue_ - mx;
+    const NumericalScalar sk2 = metaModelResult_.getConditionalCovariance(x)(0, 0);
+    const NumericalScalar sk = sqrt(sk2);
+    if (!SpecFunc::IsNormal(sk)) return NumericalPoint(1, -SpecFunc::MaxNumericalScalar);
+    const NumericalScalar ratio = fmMk / sk;
+    NumericalScalar ei = fmMk * normal_.computeCDF(ratio) + sk * normal_.computePDF(ratio);
+    if (noiseModel_.getOutputDimension() == 1) // if provided
+    {
```

Comment:
Only the handling of noise would differ between the two classes, the core of the algorithm is the same. I chose to expose only one class.